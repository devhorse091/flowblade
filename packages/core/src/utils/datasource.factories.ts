import type { QMeta, QMetaSqlSpan } from '../meta/q-meta';
import { type QError, QResult } from '../query-result/q-result';

/**
 * Create a new SQL span with sql query + params and initial timeMs: 0, affectedRows: 0.
 */
export const createSqlSpan = ({
  sql,
  params,
  timeMs,
  affectedRows,
}: {
  sql: QMetaSqlSpan['sql'];
  params: QMetaSqlSpan['params'];
  timeMs?: number;
  affectedRows?: number;
}): QMetaSqlSpan => {
  return {
    type: 'sql',
    sql,
    params,
    timeMs: timeMs ?? 0,
    affectedRows: affectedRows ?? 0,
  };
};

export const createQResultSuccess = <T extends unknown[]>(
  data: T,
  meta: QMeta
): QResult<T, never> => {
  return new QResult({
    meta,
    data,
  });
};

export const createQResultError = (
  error: QError,
  meta: QMeta
): QResult<never, QError> => {
  return new QResult({
    meta,
    error,
  });
};
