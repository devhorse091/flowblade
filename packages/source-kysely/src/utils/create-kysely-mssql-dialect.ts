import { MssqlDialect } from 'kysely';
import * as tarn from 'tarn';
import * as tedious from 'tedious';

/**
 * Create a Kysely dialect for Microsoft SQL Server.
 *
 * @example
 * ```typescript
 * const jdbcDsn = "sqlserver://localhost:1433;database=db;user=sa;password=pwd;trustServerCertificate=true;encrypt=false";
 * const tediousConfig = TediousConnUtils.fromJdbcDsn(jdbcDsn);
 * const tediousConnection = new Tedious.Connection(tediousConfig);
 *
 * const dialect = createKyselyMssqlDialect(tediousConfig, {
 *   // Optional tarn pool options
 *   tarnPool: {
 *     min: 0,
 *     max: 10
 *   }
 * });
 *
 * const db = new Kysely<DB>({
 *   dialect
 * })
 * ```
 */
export const createKyselyMssqlDialect = (
  tediousConfig: tedious.ConnectionConfiguration,
  options?: {
    tarnPool?: {
      /** default: 0 */
      min?: number;
      /** default: 10 */
      max?: number;
    };
  }
) => {
  const { tarnPool = {} } = options ?? {};
  const { min = 0, max = 10 } = tarnPool;
  return new MssqlDialect({
    tarn: {
      ...tarn,
      options: {
        min,
        max,
      },
      /**
       * @todo when https://github.com/kysely-org/kysely/pull/1073/files is merged
       * Controls whether connections are validated before being acquired from the pool. Connection validation performs additional requests to the database.
       *
       * Defaults to `true`.
       */
      // validateConnections: false
    },
    tedious: {
      ...tedious,
      // See https://github.com/kysely-org/kysely/issues/1161#issuecomment-2384539764
      TYPES: { ...tedious.TYPES, NVarChar: tedious.TYPES.VarChar },
      connectionFactory: () => {
        return new tedious.Connection(tediousConfig);
      },
      /**
       * @todo when https://github.com/kysely-org/kysely/pull/1073/files is merged
       *
       * Controls whether connections are reset to their initial states when released back to the pool. Resetting a connection performs additional requests to the database.
       * See {@link https://tediousjs.github.io/tedious/api-connection.html#function_reset | connection.reset}.
       *
       * Defaults to `true`.
       */
      // resetConnectionOnRelease: false
    },
  });
};
