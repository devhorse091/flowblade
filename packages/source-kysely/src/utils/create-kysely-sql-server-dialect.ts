import { MssqlDialect } from 'kysely';
import * as tarn from 'tarn';
import * as Tedious from 'tedious';

/**
 * Create a Kysely dialect for Microsoft SQL Server.
 *
 * @example
 * ```typescript
 * import * as Tedious from 'tedious';
 * import { TediousConnUtils, createKyselySqlServerDialect } from '@flowblade/source-kysely';
 *
 * const jdbcDsn = "sqlserver://localhost:1433;database=db;user=sa;password=pwd;trustServerCertificate=true;encrypt=false";
 * const tediousConfig = TediousConnUtils.fromJdbcDsn(jdbcDsn);
 * const tediousConnection = new Tedious.Connection(tediousConfig);
 *
 * const dialect = createKyselySqlServerDialect(tediousConfig, {
 *   // Optional tarn pool options
 *   tarnPool: {
 *     min: 0,
 *     max: 10
 *   },
 *   // Optional: customize tedious types
 *   // Example based on https://github.com/kysely-org/kysely/issues/1161#issuecomment-2384539764
 *   tediousTypes: { ...Tedious.TYPES, NVarChar: Tedious.TYPES.VarChar}
 * });
 *
 * const db = new Kysely<DB>({
 *   dialect
 * })
 * ```
 */
export const createKyselySqlServerDialect = (
  tediousConfig: Tedious.ConnectionConfiguration,
  options?: {
    tarnPool?: {
      /** default: 0 */
      min?: number;
      /** default: 10 */
      max?: number;
    };
    tediousTypes?: typeof Tedious.TYPES;
  }
): MssqlDialect => {
  const { tarnPool = {}, tediousTypes } = options ?? {};
  const { min = 0, max = 10 } = tarnPool;
  return new MssqlDialect({
    tarn: {
      ...tarn,
      options: {
        min,
        max,
      },
      /**
       * @todo when https://github.com/kysely-org/kysely/pull/1073/files is merged
       * Controls whether connections are validated before being acquired from the pool. Connection validation performs additional requests to the database.
       *
       * Defaults to `true`.
       */
      // validateConnections: false
    },
    tedious: {
      ...Tedious,
      // See https://github.com/kysely-org/kysely/issues/1161#issuecomment-2384539764
      ...(tediousTypes === undefined ? {} : { TYPES: tediousTypes }),
      connectionFactory: () => {
        return new Tedious.Connection(tediousConfig);
      },
      /**
       * @todo when https://github.com/kysely-org/kysely/pull/1073/files is merged
       *
       * Controls whether connections are reset to their initial states when released back to the pool. Resetting a connection performs additional requests to the database.
       * See {@link https://tediousjs.github.io/tedious/api-connection.html#function_reset | connection.reset}.
       *
       * Defaults to `true`.
       */
      // resetConnectionOnRelease: false
    },
  });
};
