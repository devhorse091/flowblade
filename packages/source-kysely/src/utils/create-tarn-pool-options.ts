const THIRTY_SECONDS_IN_MS = 30_000;

export type TarnPoolOptions = {
  /** Minimum number of connections, default 0 */
  min?: number;
  /** Maximum number of connections, default 10 */
  max?: number;

  /**
   * The maximum number of milliseconds to wait for a connection to become available when the pool is exhausted.
   * default: 30_000 (30 seconds)
   */
  acquireTimeoutMillis?: number;
  /**
   * The maximum number of milliseconds to wait for a connection to be created.
   * default: 30_000 (30 seconds)
   */
  createTimeoutMillis?: number;
  /**
   * The maximum number of milliseconds a connection can go unused before it is closed.
   * default: 5_000 (5 seconds)
   */
  destroyTimeoutMillis?: number;
  /**
   * The maximum number of milliseconds that a connection can remain idle in the pool before being closed.
   * default: 30_000 (30 seconds)
   */
  idleTimeoutMillis?: number;
  /**
   * The frequency to check for idle connections.
   * default: 1_000 (1 second)
   */
  reapIntervalMillis?: number;
  /**
   * How long to idle after failed create before trying again
   * default: 200
   */
  createRetryIntervalMillis?: number;
  /**
   * Whether to propagate errors that occur during connection creation.
   * default: false
   */
  propagateCreateError?: boolean;
};

const defaultPoolOptions = {
  min: 0,
  max: 10,
  acquireTimeoutMillis: THIRTY_SECONDS_IN_MS,
  createTimeoutMillis: THIRTY_SECONDS_IN_MS,
  destroyTimeoutMillis: 5000,
  idleTimeoutMillis: THIRTY_SECONDS_IN_MS,
  reapIntervalMillis: 1000,
  createRetryIntervalMillis: 200,
  propagateCreateError: false,
} as const satisfies TarnPoolOptions;

export const createTarnPoolOptions = (
  tarnPoolOptions?: TarnPoolOptions
): Required<TarnPoolOptions> => {
  return {
    ...defaultPoolOptions,
    ...tarnPoolOptions,
  };
};
