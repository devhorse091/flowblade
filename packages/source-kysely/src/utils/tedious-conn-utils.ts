import { convertJdbcToDsn, parseDsnOrThrow } from '@httpx/dsn-parser';
import type { ConnectionConfiguration as TediousConnectionConfig } from 'tedious';
import * as v from 'valibot';

type AuthConnectionOptions = NonNullable<
  TediousConnectionConfig['authentication']
>['options'];

const tediousSchema = v.object({
  database: v.string(),
  authentication: v.optional(
    /**
     * Different options for authentication types:
     *
     * `default`: [[DefaultAuthentication.options]]
     * `ntlm` :[[NtlmAuthentication]]
     * `token-credential`: [[CredentialChainAuthentication.options]]
     * `azure-active-directory-password` : [[AzureActiveDirectoryPasswordAuthentication.options]]
     * `azure-active-directory-access-token` : [[AzureActiveDirectoryAccessTokenAuthentication.options]]
     * `azure-active-directory-msi-vm` : [[AzureActiveDirectoryMsiVmAuthentication.options]]
     * `azure-active-directory-msi-app-service` : [[AzureActiveDirectoryMsiAppServiceAuthentication.options]]
     * `azure-active-directory-service-principal-secret` : [[AzureActiveDirectoryServicePrincipalSecret.options]]
     * `azure-active-directory-default` : [[AzureActiveDirectoryDefaultAuthentication.options]]
     */
    v.union([
      v.literal('default'),
      v.literal('ntlm'),
      v.literal('token-credential'),
      v.literal('azure-active-directory-default'),
      v.literal('azure-active-directory-password'),
      v.literal('azure-active-directory-access-token'),
      v.literal('azure-active-directory-msi-vm'),
      v.literal('azure-active-directory-msi-app-service'),
      v.literal('azure-active-directory-service-principal-secret'),
      v.literal('azure-active-directory-default'),
    ]),
    'default'
  ),
  user: v.optional(v.string()),
  password: v.optional(v.string()),
  trustServerCertificate: v.optional(v.boolean(), false),
  clientId: v.optional(v.string()),
  useUtc: v.optional(
    v.boolean('Use UTC time zone for dates (default: true)'),
    true
  ),
  encrypt: v.optional(v.boolean(), true),
  connectTimeout: v.optional(v.number('Number in milliseconds.')),
  requestTimeout: v.optional(v.number('Request timeout in milliseconds')),
});

export const TediousConnUtils = {
  /**
   * Parse and validate a JDBC connection string and return a Tedious connection configuration.
   *
   * @example
   * ```typescript
   * const jdbcDsn = "sqlserver://localhost:1433;database=db;user=sa;password=pwd;trustServerCertificate=true;encrypt=false";
   * const tediousConfig = TediousConnUtils.fromJdbcDsn(jdbcDsn);
   * const tediousConnection = new Tedious.Connection(tediousConfig);
   * ```
   * @throw TypeError if dsn isn't valid
   */
  fromJdbcDsn: (jdbcUrl: string): TediousConnectionConfig => {
    const dsn = convertJdbcToDsn(jdbcUrl);
    const parsed = parseDsnOrThrow(dsn);
    const params = v.parse(tediousSchema, parsed.params);
    const {
      authentication,
      database,
      useUtc,
      encrypt,
      connectTimeout,
      trustServerCertificate,
      clientId,
      user,
      password,
      requestTimeout,
    } = params;
    let authConnOptions: AuthConnectionOptions;
    switch (authentication) {
      case 'default':
        authConnOptions = {
          userName: user,
          password: password,
        };
        break;
      case 'azure-active-directory-default':
      case 'azure-active-directory-msi-app-service':
        authConnOptions = {
          clientId: clientId,
        };
        break;
      default:
        throw new Error(`Unsupported authentication type: ${authentication}`);
    }
    return {
      server: parsed.host,
      authentication: {
        type: authentication,
        // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
        options: authConnOptions,
      },
      options: {
        database: database,
        port: parsed.port,
        useUTC: useUtc,
        encrypt: encrypt,
        trustServerCertificate,
        ...(requestTimeout ? { requestTimeout } : {}),
        ...(connectTimeout ? { connectTimeout } : {}),
      },
    };
  },
};
