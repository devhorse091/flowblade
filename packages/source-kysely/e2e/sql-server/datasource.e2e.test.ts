import { sql } from 'kysely';

import { createE2eDatasource } from '../utils/create-e2e-datasource';

describe('Datasource sqlserver', () => {
  const ds = createE2eDatasource('sql-server');
  describe('Kysely raw queries', () => {
    it('01. basicQuery', async () => {
      type Row = {
        one: number;
      };

      const rawSql = sql<Row>`SELECT 1 as one`;
      const result = await ds.queryRaw(rawSql);
      const stabletimeMs = 0.1;
      result.meta!.timeMs = stabletimeMs;
      expect(result).toStrictEqual({
        data: [
          {
            one: 1,
          },
        ],
        meta: {
          affectedRows: 1,
          timeMs: stabletimeMs,
          query: {
            params: [],
            sql: 'SELECT 1 as one',
          },
        },
      });
      assertType<Row[]>(result.data);
    });

    it('02. basicQuery with params', async () => {
      type Row = {
        one: number;
      };
      const params = {
        number: 1,
        string: 'Hello',
      };

      const rawSql = sql<Row>`
          SELECT 1 as one
          WHERE 1 = ${params.number}
          AND 'Hello' like ${params.string}
      `;
      const rows = await ds.queryRaw(rawSql);
      const stabletimeMs = 0.1;
      rows.meta!.timeMs = stabletimeMs;
      expect(rows).toMatchSnapshot();
    });
  });

  describe.skip('Kysely select queries', () => {
    it('upsert some brand', async () => {
      const params = [
        { name: 'Brand A' },
        { name: 'Brand B' },
        { name: 'Brand C' },
      ];
      const qRaw = sql`
        DELETE FROM brand WHERE 1=1;    
        INSERT INTO brand (name, created_at) VALUES (
            SELECT name, CURRENT_TIMESTAMP
            FROM OPENJSON(${JSON.stringify(params)}) WITH (
                name NVARCHAR(255)
            )
        )      
      `;
      expect(await ds.queryRaw(qRaw)).toMatchSnapshot();
    });

    it('get some brands', async () => {
      const query = ds
        .eb()
        .selectFrom('brand as b')
        .select(['b.id', 'b.name'])
        .leftJoin('product as p', 'p.brand_id', 'b.id')
        .select(['p.id as product_id', 'p.name as product_name'])
        .where('b.created_at', '<', new Date())
        .orderBy('b.name', 'desc');

      const rows = await ds.query(query);
      const stabletimeMs = 0.1;
      rows.meta!.timeMs = stabletimeMs;
      expect(rows).toMatchSnapshot();
    });
  });
});
