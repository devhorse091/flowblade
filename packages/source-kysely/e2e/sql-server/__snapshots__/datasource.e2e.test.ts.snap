// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Datasource sqlserver > Kysely raw queries > 02. basicQuery with params 1`] = `
{
  "data": [
    {
      "one": 1,
    },
  ],
  "meta": {
    "affectedRows": 1,
    "query": {
      "params": [
        1,
        "Hello",
      ],
      "sql": "
          SELECT 1 as one
          WHERE 1 = @1
          AND 'Hello' like @2
      ",
    },
    "timeMs": 0.1,
  },
}
`;

exports[`Datasource sqlserver > Kysely select queries > basicQuery with params 1`] = `
{
  "data": [],
  "meta": {
    "affectedRows": 0,
    "query": {
      "params": [],
      "sql": "select "b"."id", "b"."name" from "brand" as "b"",
    },
    "timeMs": 0.1,
  },
}
`;
